example/api.js

```diff
@@ -0,0 +1,107 @@
+async function sha256(message) {
+  const msgBuffer = new TextEncoder().encode(message);
+  const hashBuffer = await crypto.subtle.digest('SHA-256', msgBuffer);
+  return u8aToHex(new Uint8Array(hashBuffer));
+}
+
+// Create Passkey Challenge
+export async function createPasskeyChallenge(
+  frequencyUrl,
+  accountPublicKey,
+  nonce,
+  accountOwnershipProof,
+  balanceTransferTxHash,
+) {
+  const api = await loadApi(frequencyUrl);
+  const call = hexToU8a(balanceTransferTxHash);
+  const ext_call_type = api.registry.createType('Call', call);
+  const passkeyCall = {
+    accountId: accountPublicKey,
+    accountNonce: nonce,
+    accountOwnershipProof: {
+      Sr25519: accountOwnershipProof,
+    },
+    call: ext_call_type,
+  };
+  const passkeyCallType = api.createType('PalletPasskeyPasskeyCall', passkeyCall);
+  const calculatedChallenge = await sha256(passkeyCallType.toU8a());
+  const challenge = {
+    challenge: calculatedChallenge,
+    passkeyCall: u8aToHex(passkeyCallType.toU8a()),
+  };
+  return challenge;
+}
+
+// Base64 URL to Uint8Array Conversion
+export const base64UrlToUint8Array = (base64Url) => {
+  // Decode base64 URL encoding to base64
+  const base64String = base64Url.replace(/-/g, '+').replace(/_/g, '/');
+  // Convert to Uint8Array
+  const binaryString = atob(base64String);
+  const len = binaryString.length;
+  const bytes = new Uint8Array(len);
+  for (let i = 0; i < len; i++) {
+    bytes[i] = binaryString.charCodeAt(i);
+  }
+  return bytes;
+};
+
+// Create Passkey Transaction
+export async function sendPasskeyTransaction(
+  frequencyUrl,
+  passkeyPublicKey,
+  originalChallenge,
+  passkeySignature,
+  passkeyAuthenticatorData,
+  passkeyClientData,
+  passkeyCall,
+) {
+  const api = await loadApi(frequencyUrl);
+  const call = hexToU8a(passkeyCall);
+  const passkeyCallType = api.createType('PalletPasskeyPasskeyCall', call);
+  const passKeyPublicKeyBytes = hexToU8a(passkeyPublicKey);
+  const passkeySignatureBytes = hexToU8a(passkeySignature);
+  const challengeReplacer = '#rplc#';
+
+  let clientDataRaw = base64UrlToUint8Array(passkeyClientData);
+  let authenticatorData = base64UrlToUint8Array(passkeyAuthenticatorData);
+
+  // Convert clientDataRaw to a string, replace challenge, and then convert back to Uint8Array
+  const clientData = new TextDecoder().decode(clientDataRaw).replace(originalChallenge, challengeReplacer);
+  const clientDataBytes = new TextEncoder().encode(clientData);
+
+  const passkeyTransactionPayload = {
+    passkeyPublicKey: Array.from(passKeyPublicKeyBytes),
+    verifiablePasskeySignature: {
+      signature: Array.from(passkeySignatureBytes),
+      authenticatorData: Array.from(authenticatorData),
+      clientDataJson: Array.from(clientDataBytes),
+    },
+    passkeyCall: passkeyCallType,
+  };
+
+  const payload = api.createType('PalletPasskeyPasskeyPayload', passkeyTransactionPayload);
+  const tx = api.tx.passkey.passkey(payload);
+  return tx.toHex();
```

Comment:
current it is just returning encoded tx that can be sent to chain whenever.

Question: Do we send the transaction in example given example requires frequency url for transactions

Reply:
I really like the idea of just showing the hex.

I think it makes a few things simpler and gives us some options:

- Clearer name: "Generate Transaction" instead of "send transaction"
- Instead of asking for the transaction hash in step 4, we can just generate a "standard" transaction that sends 1 token to alice `5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY`. (Because it doesn't matter if it is able to be executed)
- We could also link directly to the decode transaction option in Polkadotjs App like [this](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2F0.rpc.testnet.amplica.io#/extrinsics/decode/0xcd5704010091034f024808949b660e6f7787eb7f1337b51258aa17dcbbdf7b8c43d10b24712e6722463e00481039c790e4209a63036b4afee4c0823b318bb685042ad29e21f50b4bbcf4227b606310bb0fd08a89326a83056f012d1d2a31345d6346038301298c600286b70c0661757261200cf38f0800000000045250535290b8e1410d295797489ae48957256463827359f0af3d25b50a8a3fdd5fbe1e4e1216aa9a00056175726101016cab58de63da49a0a29b7357b59eb33b76d38496d17dd429a8d2185416224651f5c47967e03e248dccc6d42d7a258556782cefdcb97965ef5b7bbcb46527e08c87aa260081e7a9776a5e9a997277cb6b5f5ed84f9f24adb7c078c15eb026bbc1f42c133b000050008ca5030000300000800000aaaa020000001000fbff00000500000005000000020000000200000003000000020000000000500000c800001e0000000000000000000000000000000000000000000000000000000000000000000000e8030000009001001e000000009001000c01002000000600c4090000000000000601983a000000000000b004000000000000000000001027000080b2e60e80c3c90180969800000000000000000000000000050000001400000004000000020000000000060000006400000002000000190000000000000002000000020000000200000005000000020000000001000000840103e757b3b044921698db0370bd579088833e08331aece49f5460b4cabac6b5afb03630751aa5440ab4a00f00000f0ec7912296d6e7733de0d882f466a595cde5d5f324ad09644b4ecf84c2fab8c03e77dfdb8adb10f78f10a5df8742c545821d41eedd87367c4022e348017e4b8f3ca6b2eca49f2bebbb4390ba0defb15bc83f0004b49d95320d9021994c850f25b8e385cec92c0f4e04b1016027e6db18e8a89778502819dac74eaf4b1427f518b4f08eb8570c58f3b5989bb72cdf07000080a4ad8533d3dc829c365c264ebe8a9541e269fdca7e1ff34efc195e19a618a0e6c45e414cb008e0e61e46722aa60abdd672807a4fbb241424af6d8f1c5de69f353c07fc744bdbd4e50db9bf5570c9612cdf66c45ee678799d3eff024253b90e84927cc680b815c122a7b1cf44849f23798abff04a35f8a040fe84d2963be473553d6394ad150180001480473b55a9bdc9a87e0a4ee9f9edf43f10656aec12f46ee761d3d9f45cf2c62ca68007d4ff6c1e1c1e6f84b2755f9387fef6940f3feea49b6909b612195039b9b8bee080006080af1466f78b5326632ba5476bf69784516d05ac110e53bd2a145fd2e26506e7b34c5e7b9012096b41c4eb3aaf947f6ea429080000150180011080346fdc0e599ce19761a88ceaf1f19f65ebb7c1ec24961cbca6af25a97027a151800ebe650369ce0e3ce85b22816181f598cc6ecf7a3b22314357ab4181b9ccb86715018001108070ac9add9cf11a04ed6a184e4090995d25190ecb5c2e7d3c5a21f6452577fad4806080a0756113eb8ee2980e43ba6f8ddd521ffef0a157a1eaa6b740895924626c1d02800219803936e1cb6f892433d2fcc4d289dcbb1a4cde45dd53e19ae5a3e641735cf083e980716f892eead010b2d19aff4c72f44eba1bf6c51a6ee81c26d66c387c5894e092804442e47b9281b6ef62450c2ed816fceacb242d21783bb312cb08be917aaf635e801769f9cf0e69001e5c69f1f3fd66d28f3bf20a1a1fd28e92f406e0cc3f734e2c1d028002a480d6271a4c2a039dde0fa546dc399ed48e95a8fb69893042beeea184dd1d9f3b488012f25b456f49cd0dc37bb71cc74a29c7a88065189fe13f9a147e12f0a4db40098010f31418fee338aacf8bbf564ed8bef2ecf0c69bdd1b287567863b2d720dcd20800d91452f876bfdbee898d18214931022c221447d69322f34204a7aef6cf11eea1501800404804c5b5b4737f7968eac0faefa5e23f5914741c70524f21ee6ae47a3de2a57a9f780ee3cefde9becc4b979a6a8b0f26e44656fb9640467a214f7f980bcae7d338996dc800820800e8ca0cb46a5deb990ac9b0b66f73ecc56f4057602b34a9c889a9b9e62a116a8485ead6eef5c4b1c68eaa71ea17a02d9de0400a10280101780cbe9c011576212a8e529810612b4f50780d1297bdb2203449f6a83b7e3e6639380201ddbd791ebc350ff9847921218fe0e3f93add9eac7de00f404da527d25445380c3f4ddf4e42802b5bfc0d2afd8d37f63f96f4a5e3d2b67eeffcb8cfce72fd35c80a2e6df0a09d1a9d999383287f3956505a785f4d8afa869091f4fff028700e74480bdd0c55bbcd9429e4d1f24255e34ff9658fd9cb2f7e3ba7b55147cb6735e7557a1028014648090e63f0626f0b1b06b2aa6a1c9708679ddd2fcba0ebd655b51770b3b925592908063fa3443f4e8b71a14886166ad03f5747914d180a4f9766ef2f500531e14c60880f6f6801e4b41e2e6d8ec194dba122bfb9eb33feb2545ef5144cea79551f7cc52800b9208ec08029f2b4f66ec1b95ec78e6d1bd4eb732a357b657de63371cc61ba0808c0cd1b46668a6d6451e533bbc78b0980d581096940c68f0b26014c0ca41b3a915018021008065ba2334525527aad98b48cd5e556ad718181ce6018c6bd459abf0b0be185d80806008fcee1f70a1e1225fa2345b93f9fdaf2ee822d4493d844b566f3706f15e3e1d028088058010762505ef650cac2df7e22f14aead5fa62992884ccbe5f70e3b4fe459bd1e4a80174ca65ec3aae4ad3e40ed40c654a04e5f2cfdc89483cda40a36b0b858ba6b8d8093617f28e050965a14e7f7a625faf4a2fc1887d468374e9be51b794c0adfd8b7805e29eda83ab444479ac20e2c6609cc335b07d3d17b790042f11782fda16fe386150180c000809418c0b5d5f4ce95d1780e7cbd3856b3a05c2fda8f7864e4d4bba4e4e1cbf6c780e05696bf5fed6f6fe509849e9fc2936a6fe1c80f204ab5fb8bfaf83412808994a10280e081807d15ea5039b964b37b7d540bdcd7d21c1f704aaca136ea7a0d8a265e3042618780305129439f578f736467197c47fa470a87f84a59e319c6752e88532057c1aed48044a8ad9461d0587a927b8f78fbd9ede6ee6fc0e86469a9def63473a2f73c0847804b05e58faaf56eb840e0e605857f2c72b2c5b3b4f3a9ca98c980455e0219058f806e50324cad399ba8743c9c879fbe91f2be7cbc551211111ec78733a611924e194d0880ffff80786bed85c6e58b66d60d1802184917dae3fe733899255b59eb28b5f304074ad58067c7f9969b0a8cccba69baa449464d1fcf171da3ebe7e4f555270a3b26c6853f80e4460b63ac5a9d70b2cf8b37d15e3d81ec7cee235b640ab30b61d7e0061c5f1580479fe9a4a29a588a6f7c5787908789bc43729b8e527d4b803a3f821e16d52b698023c1ac3d3306183ba7ee94654054ccce59c5f43a7cf5ba5b52d2cc17f1dadcc18007a6e93f424f185f91ac1cb162beb171152b70c6d50c53f86f34fb9f4f919dd080b63eeb81fc92a01e51e5e44ecf527e82596ecbb88922752eaca0663e5feca01980c9a07f6203223cc27859417fa82bb49f5533082dc27d96e60bc6ab894946b70780e84f6ea0e76b056478112a8dea96f7f89172956a098f8c43a89d22c7ff47f51880d746c5e1019d4e2407076de3cae4a3c10e870737565d143c96ec7865efd2c3ec802dcd772353472f197ed41f9cd920d46791b866f496f299dedb3c843a3168a510800c9cf6139c57bdf550ff58341990b9148e24d17ea8b420aeaa6395281fa6fc5e8028f1f28a856b76a76594a9b0e83770dd5d7e0a6b781a90264c8e9b7c31965682806af5ef2e45ca57748c414dc5d830845e59b2f825ea75f7f4261b291e6aba543a80b2ec7bbfd0f12271feeac94618efb392770bb561aa61efb942ba8915b484aaf68026dcf0b26daf4939df319bfc6589bc3acce7b03812993ef96434fa3c403b8b20990391034f024808949b660e6f7787eb7f1337b51258aa17dcbbdf7b8c43d10b24712e6722463e00481039c790e4209a63036b4afee4c0823b318bb685042ad29e21f50b4bbcf4227b606310bb0fd08a89326a83056f012d1d2a31345d6346038301298c600286b70c0661757261200cf38f0800000000045250535290b8e1410d295797489ae48957256463827359f0af3d25b50a8a3fdd5fbe1e4e1216aa9a00056175726101016cab58de63da49a0a29b7357b59eb33b76d38496d17dd429a8d2185416224651f5c47967e03e248dccc6d42d7a258556782cefdcb97965ef5b7bbcb46527e08ca9049d0da05ca59913bc38a8630590f2627c17d9806ad041b61bad6d6cca7841efe1e07affcddada1338457ec724262dfa8f406aa580592c6b0b62e04550afaa4aecf62a63690514a4a2b35a682439a6d0e5790cd73a6c5f0a351b6a99a5b21324516e668bb86a572404290d0000f2070000505f0e7b9012096b41c4eb3aaf947f6ea4290800008073a3fda0a1edb6b960cf130de178c39b769db225a557d67cf6ab79ca81e413f28012f13ceb83b98d0ed867bbfa8f68432e90d4328639e22f32b661873a06c24d0c80bca5ba0d9f870b25e86b03a61278ac34e75dda4151431770fa0044c30d9e57fa8028e77b9e91cccedbc435f20b2caee5fcbf0d47ff2f1d47745e1638db1278e55280825eba4d89ccfe333512fd9465c352d79d66f3509774b2ad7d2bb759fce7c960d9049e710b30bd2eab0352ddcc26417aa1945fc380a15899d9e670406c9fc53eabc99c138b721c1962971eef2a1edb5822bd10bd6b8037d5193550825cc30775e9368b9d892ca03d7b26ad45bb584e6ff5120bf0f68480975dcf5a6c2e977deebbd96b655e786894127c6509c3cbc1521d9a3ced73b5d780a1ba18f1b041970114a42d976b1901321735389a756c973388f2b8be06e21b0a505f0e7b9012096b41c4eb3aaf947f6ea4290800004c5f03c716fb8fff3de61a883bb76adb34a204008006b5738022c31086fa4d7c89c850d43eab57d057050d2eda46e5a63ebd42a4694c5f0f4993f016e2d2f8e5f43be7bb25948604008022522772fdcfc5b8dd57d209e2ea03883eb85894fb30dd07c6074b74efb887f0809ea3f10744bc47084d80b97e9e57d6767b36a67ef1fdaf36d416de5c554bf04271059e7fefc408aac59dbfe80a72ac8e3ce5d0bf80951b05627be9d7d110f7239f65aecf81d33056fa77109b37fa187bad18ab8bae8038a96110b1e66b070cc2f67705c9be03f5a9fb14d59a440ba45ca3d5d3707caf808dd8ef007f5bc3b2a593ed5f4a26434ca3cf4240530a168c8d7be88993c8d717803404e7007153a4cbc8e1b6bee97c8db0a53a52da6f4a4c4fa67f06916f755ff7802523c9854c05aab97f613f60222a60e5becc59f11544748f998a9c8203fac0288064a0aadad84352461f321a34f1d88a8a1f52c05268dbfc3e0a739f1049d921a380ab736ba1915d465798b18a9bd65513f086f33901c68470dff64c539a49b52ac5807ee20bc5d90115fd156d28cdec230108c0233ade3d13e271268a341392f8720780c0f6890f44443c8dd19ea30df7d12be39755ea05c2f14dd8a3f03eca2dea70c4804605a5942feea1e3be7ce112be7b079818cfb7e249bad23d380230a0638010f2f9049eb6f36e027abb2091cfb5110ab5087ff96c685f06155b3cd9a8c9e5e9a23fd5dc13a5ed201ae61f1100000000685f08316cbf8fa0da822a20ac1c55bf1be320c010000000000000505f0e7b9012096b41c4eb3aaf947f6ea42908000080495281e395eecaa76c1c20078d9aa05c58b0d82aaaa32148174d99f21aca24f780c6c4d708468891f3ab25e5ca8ede764021417f794f3d9288fec3dfd61ea05333806beed985f83da12e99f64c2dc371aa4d7d615bf5f99861413e4e8c0ed78cf57480e0408168890d13c4ba2ea2e9fd775531edbc61182adae8293ecdff727bc04e0080b04e2a20e3ecdb1d41359a3eea0e3d572b021d6d970cae0a5ca524a2217053c380eed21cade6ad1dddad68364de90343d7202bd862ac082dbc17b6e3f6f6ab80f0685f090e2fbf2d792cb324bffa9427fe1f0e20eea7260045aa260071019ede3d8a54d27e44a9d5ce189618f22d3008505f0e7b9012096b41c4eb3aaf947f6ea429080b004c5f03b4123b2e186e07fb7bad5dda5f55c0040080d250445a188d62af2408042105c2a18c0263e0dca25254fb2a8a49ccabd4f609d5019ef78c98723ddc9073523ef3beefda0c104480ed77da9028da4587e6197bfa7fd99045f8d8d930b0b69125717ca34bd872020c80fb74edb3b8be1729eb1597878c4815e7860772f68220a7448d2df3ce1bc5883880d9adb8664b679c2cacbb0e7f5d26f6d2ca643cb85b454245e4ccecd58a8d6b5a61029f012b746dcf32e843354583c9702cc020508d4c570e26472d4991124a290d00001404e80300004c5707270545ab1f73a4390800001404e80300004c570c58f3b5989bb72cdf0700001404e80300004c570d70e225a2d11feea10f00001404e803000080b5ddc825db2f247c3ad2f516f9b6f4008ec51c0c98cbb732af1a3686a81643154c570d3c4eac5d2f7717dd0700001404e803000085079f0b3c252fcb29d88eff4f3de5de4476c3dfbf8081b549cc652a2519341277a4c0f1685b4585df8f8261abcf1880a71e3bbd49f48075d490ecd36631ad094b9a3cc33b1300264ae9980b984680af278cbad24fa0408097cb9286e595ad539658087e95392f9ba1492986c8c8d427d97561ebbccf2d4580d4d4a45e9c1df804aa2bbc74c56a50c3eabb973b416ebaa5e3ba95c96b3aef24804932cf9d7974daa89112a442e6d840ad21ca268e5d70ad0ad0b947664b191d638062b35ac4d7fa93627bfa8d49c2b64a3d634d916a953dc410eea5248c3856845880bbdee6c7cf086acc345d52e3bf009079a6a53349d916ef8ba703d82829f030378013ba04d61fa0686750e7fd98d82a8a47aa4b24283e56a340634726de81765ee080c9135de74b1202b153e19707423109aea510f78cf092f63d0dfb9e613986b43c80064b5c7bb26b7bd8dcb2348ce9fe93434c7d11f1a64bc06e9856fdf88585700680318429abbb6457f279dc3050dd8b93e759a622a93e89f95e76fefb3e27ba63d78076c73cd6de7607da93b518f0d8c38cebcf3cb70c18daffb86d0d5c27d9f4650280e431c9cd20c1645496c04a745918cf1a6707432f9b7fff41736fa9257d345326809ecd4e7fac7fcfc7b6966a3ed283e7eacc4165aa428e521e0cda415470252a3a61029f0d3719f5b0b12c7105c073c507445948508d4c570e26472d4991124a290d00001404e80300004c5707270545ab1f73a4390800001404e80300004c570c58f3b5989bb72cdf0700001404e80300004c570d70e225a2d11feea10f00001404e803000080b5ddc825db2f247c3ad2f516f9b6f4008ec51c0c98cbb732af1a3686a81643154c570d3c4eac5d2f7717dd0700001404e80300005502bd007f03cfdce586301014700e2c2593fad157e461d71fd4c1f936839a5f1f3ed00958570e26472d4991124a290d0000200000000000000000585707270545ab1f73a439080000200000000000000000585703f5ad4e751c9893a90f000020000000000000000058570c58f3b5989bb72cdf070000200000000000000000585706ff6f7d467b87a9e8030000200000000000000000d901bf0e02656c61795f64697370617463685f71756575655f72656d61696e696e675f63617061636974790c643847090d000020aaaa0200000010003847090800002008000000000010003847090f000020080000000000100038470f07000020080000000000100038470803000020aaaa0200000010000000) for testnet

Reply:
Yes, renaming :+1:   added a default value to tx hash and to confirm the hex generated can be pasted in Polkadotjs decode section

Reply:
an example url

[this](https://polkadot.js.org/apps/?rpc=ws%3A%2F%2F127.0.0.1%3A9944#/extrinsics/decode/0x5506044300020485a129b9b1c6d9af157227e6c179267adea0e71c1f42f3e783129983d7ad0a010100b04e71766f2826016ebee2c1c3173d587fd3e8b9993cd861834d82f2bc1a737d90ffca8e8ad609986664aca9dc93cf0c70d99ad747e32527154e444f9d551d94589f094cd6e2bd35a7fb8df7b9bb35e3c03b1205a8c22e1201c6018cb59fa3511d00000000e5017b2274797065223a22776562617574686e2e676574222c226368616c6c656e6765223a222372706c6323222c226f726967696e223a2268747470733a2f2f706173736b65792e616d706c6963612e696f3a38303830222c2263726f73734f726967696e223a66616c73652c22616c67223a224853323536227d7e6547040af11583d554afec92d5a7b812bfcc761846e9cd4ba83c3e718174000000000001c8592e4e150defd4f49f5cc5eeb4bfb08d4964a489a7d2a071184058a71a712d01654014b76c88d23115b6b9cf012071c6f8ead3e930a1117e1b14f218a7a2850a0000d43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d070010a5d4e8)